### DISCUSSION TITLE
How to maintain the previous scroll position when clicking on a back link?

### DISCUSSION THREAD

**Discussion by thanhtran-git** [neuefische/questions#809] [Web]
[Update 12.02.24: added one more attempt of solving: 3. next.config.js]

## Good morning,

For a good user experience, I would like to keep the previous scroll position when a back link is clicked.
This avoids the user having to scroll from the top all the way down to their previous scroll position. Unfortunately, the solutions I've tried have not worked out.

## Steps to reproduce
1. Link to our project: [https://capstone-event-app.vercel.app/](https://capstone-event-app.vercel.app/)
2. Scroll down to one of the lower event cards
3. Clicking on one of them leads to the details route and card
4. Clicking on the Back Link leads to the very top home page, instead of the previous scroll position, where the card was clicked.

## I have tried the following: (Note: this code is not one the main, but still in a local test branch)

**1.)** Giving the Link a scroll={false} attribute as described in the [next docs](https://nextjs.org/docs/pages/api-reference/components/link) (section "Disable scrolling to the top of the page") to prevent the default behaviour of Links automatically scrolling to the top

`      <Link href="/" scroll={false}>
        Back
      </Link>`

**2.)** Instead of using a Link, I tried to give the same attribute to a button with a handler function to redirect me to the home page.

`      const handleClick = () => {`
`      router.push("/", undefined, { scroll: false });`

`       <button onClick={handleClick}>Back</button> `
 
**3.)** In the next.config.js, adding

  ` experimental: {
    scrollRestoration: true
  }
`
has the following effect: clicking the back button of my browser or the back button of my physical mouse indeed leads me back to the previous scroll position.
Clicking the back Link still scrolls me to the top though.

What I have also tried is, since we are working on the prehistoric version 13.4.7 of nextjs, to update to 13.5.1. Also without success.

## Question: 
Why does the scroll={false} not work as described in the docs? I don't get error message, it simply has no effect.
What are alternative solutions without having to write 50 lines of code with useEffect and useRouter? 🫠

## Searching for "scroll position" on this discussion board.................
I have found [the same issue from the user alper92](https://github.com/orgs/neuefische/discussions/696) who did not get any responses in August 2023 🥲

Thank you, good night.

**Comment by JessicaLoers** [comment]
##  1. Why does the `scroll={false}` not work as described in the docs? I don't get error message, it simply has no effect.

> Next.js Documentation [`scroll`](https://nextjs.org/docs/pages/api-reference/components/link#scroll)
Defaults to `true`. The default behavior of `<Link>` is to scroll to the top of a new route or to maintain the scroll position for backwards and forwards navigation. **When false, next/link will not scroll to the top of the page after a navigation.**


In your described goal, **where you want to return to the exact previous scroll position when a user navigates back using a "Back" link,** just setting `scroll={false}` isn't really sufficient.

This is because `scroll={false}` ...
- only prevents the page **from automatically scrolling to the top after navigation** 
- but **doesn't actively restore the previous scroll position** when a user navigates back from a detail view to a list view for example


The challenge in Next.js projects and many modern web applications is that the **scroll position is not automatically preserved across page transitions**, especially when asynchronously loaded content is involved.

💡 **This requires explicit logic to save and restore the scroll position.**

1. **Saving the Scroll Position:** Before the user leaves the page, save the current scroll position in a state that persists across page transitions, such as localStorage or sessionStorage.

2. **Restoring the Scroll Position:** When the user returns to the previous page, read the saved scroll position and set the page's scroll behavior accordingly.

## 2. What are alternative solutions without having to write 50 lines of code with useEffect and useRouter? 🫠

_You could try one of these possible solutions to choose the approach that best fits your project's and team needs, considering factors like complexity, compatibility, and specific requirements._  🙂

**1. Local Storage & Scroll Position**
Store the scroll position in localStorage or sessionStorage before navigating away, and restore it on returning to the page.

**2. [scrollIntoView ](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView) & [React useRef](https://react.dev/reference/react/useRef)**
Use React ref to target an element and scrollIntoView to scroll to it on component mount or action.

**Comment by Alexander-Iantbelidze** [comment]
maybe you install npm install use-local-storage-state, import the hook on your page and try to save the scroll position persistently in the local storage.
